File: packages/lambda/src/load_env.py
#!/usr/bin/env python3
"""
Script to load environment variables from .env files for local testing

This script loads environment variables from:
1. Root .env file
2. Lambda package .env file

Usage:
python load_env.py python local_test.py
"""

import os
import sys
import subprocess
from pathlib import Path
import re

def load_env_file(file_path):
    """Load environment variables from a .env file"""
    if not file_path.exists():
        print(f"Warning: .env file not found at {file_path}")
        return {}
    
    print(f"Loading .env file from {file_path}")
    env_vars = {}
    
    with open(file_path, 'r') as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith('#'):
                continue
                
            key, value = line.split('=', 1)
            env_vars[key] = value
    
    return env_vars

def process_env_vars(env_vars, existing_vars=None):
    """Process environment variables, resolving ${VAR} references"""
    if existing_vars is None:
        existing_vars = {}
    
    processed_vars = {}
    
    for key, value in env_vars.items():
        # Replace ${VAR} with existing_vars[VAR] or os.environ[VAR]
        def replace_var(match):
            var_name = match.group(1)
            return existing_vars.get(var_name, os.environ.get(var_name, ''))
        
        processed_value = re.sub(r'\${([^}]+)}', replace_var, value)
        processed_vars[key] = processed_value
        
    return processed_vars

def main():
    """Main function"""
    # Get paths to .env files
    root_dir = Path(__file__).resolve().parent.parent.parent.parent
    root_env_path = root_dir / '.env'
    lambda_env_path = root_dir / 'packages' / 'lambda' / '.env'
    
    # Load root .env file
    root_env_vars = load_env_file(root_env_path)
    
    # Set environment variables from root .env
    for key, value in root_env_vars.items():
        os.environ[key] = value
        print(f"Set {key}={value}")
    
    # Load lambda .env file
    lambda_env_vars = load_env_file(lambda_env_path)
    
    # Process and set environment variables from lambda .env
    processed_vars = process_env_vars(lambda_env_vars, root_env_vars)
    for key, value in processed_vars.items():
        os.environ[key] = value
        print(f"Set {key}={value}")
    
    print("Environment variables loaded successfully")
    
    # Get command line arguments (excluding script name)
    args = sys.argv[1:]
    
    if not args:
        print("No command provided")
        print("Usage: python load_env.py <command>")
        print("Example: python load_env.py python local_test.py")
        sys.exit(1)
    
    # Execute the command with the loaded environment variables
    print(f"Executing: {' '.join(args)}")
    result = subprocess.run(args, env=os.environ)
    sys.exit(result.returncode)

if __name__ == "__main__":
    main()

File: packages/frontend/load-env.js
#!/usr/bin/env node

/**
 * Script to load environment variables from .env files for frontend development
 * 
 * This script loads environment variables from:
 * 1. Root .env file
 * 2. Frontend package .env file
 * 
 * Usage:
 * node load-env.js npm start
 */

const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');
const dotenv = require('dotenv');

// Paths to .env files
const rootEnvPath = path.resolve(__dirname, '../../.env');
const frontendEnvPath = path.resolve(__dirname, './.env');

// Load environment variables from .env files
function loadEnvFiles() {
  console.log('Loading environment variables...');
  
  // Check if root .env file exists
  if (fs.existsSync(rootEnvPath)) {
    console.log(`Loading root .env from ${rootEnvPath}`);
    const rootEnv = dotenv.parse(fs.readFileSync(rootEnvPath));
    
    // Set environment variables from root .env
    Object.entries(rootEnv).forEach(([key, value]) => {
      process.env[key] = value;
      console.log(`Set ${key}=${value}`);
    });
  } else {
    console.warn('Root .env file not found');
  }
  
  // Check if frontend .env file exists
  if (fs.existsSync(frontendEnvPath)) {
    console.log(`Loading frontend .env from ${frontendEnvPath}`);
    const frontendEnv = dotenv.parse(fs.readFileSync(frontendEnvPath));
    
    // Process variables with ${VAR} syntax
    const processedEnv = {};
    Object.entries(frontendEnv).forEach(([key, value]) => {
      // Replace ${VAR} with process.env.VAR
      const processedValue = value.replace(/\${([^}]+)}/g, (_, varName) => {
        return process.env[varName] || '';
      });
      
      process.env[key] = processedValue;
      processedEnv[key] = processedValue;
      console.log(`Set ${key}=${processedValue}`);
    });
  } else {
    console.warn('Frontend .env file not found');
  }
  
  console.log('Environment variables loaded successfully');
}

// Execute the command with the loaded environment variables
function executeCommand(args) {
  console.log(`Executing: ${args.join(' ')}`);
  
  const child = spawn(args[0], args.slice(1), { 
    stdio: 'inherit',
    env: process.env
  });
  
  child.on('close', (code) => {
    process.exit(code);
  });
}

// Main function
function main() {
  // Load environment variables
  loadEnvFiles();
  
  // Get command line arguments (excluding node and script name)
  const args = process.argv.slice(2);
  
  if (args.length === 0) {
    console.error('No command provided');
    console.log('Usage: node load-env.js <command>');
    console.log('Example: node load-env.js npm start');
    process.exit(1);
  }
  
  // Execute the command
  executeCommand(args);
}

// Run the main function
main();
